image-builder/image_builder.sh

```diff
@@ -38,299 +56,373 @@ Options:
+	# rootfs start + DAX header size + rootfs end
+	local reserved_size_mb=$((rootfs_start + dax_header_sz + rootfs_end))
+
+	disk_size="$(calculate_required_disk_size "${rootfs}" "${fs_type}" "${block_size}")"
 
-	calculate_img_size
-	if [ ${OLD_IMG_SIZE} -ne 0 ]; then
-		info "Image size ${OLD_IMG_SIZE}MB too small, trying again with size ${IMG_SIZE}MB"
+	img_size="$((disk_size + reserved_size_mb))"
+	if [ -n "${root_free_space_mb}" ]; then
+		img_size="$((img_size + root_free_space_mb))"
 	fi
 
-	info "Creating raw disk with size ${IMG_SIZE}M"
-	qemu-img create -q -f raw "${IMAGE}" "${IMG_SIZE}M"
-	OK "Image file created"
+	remaining="$((img_size % mem_boundary_mb))"
+	if [ "${remaining}" != "0" ]; then
+		img_size=$((img_size + mem_boundary_mb - remaining))
+	fi
 
-	# Kata runtime expect an image with just one partition
-	# The partition is the rootfs content
+	echo "${img_size}"
+}
 
-	info "Creating partitions"
-	parted -s -a optimal "${IMAGE}" \
-		   mklabel gpt -- \
-		   mkpart primary "${FS_TYPE}" 1M -1M \
-		   print
-	OK "Partitions created"
+setup_loop_device() {
+	local image="$1"
 
 	# Get the loop device bound to the image file (requires /dev mounted in the
 	# image build system and root privileges)
-	DEVICE=$(losetup -P -f --show "${IMAGE}")
+	device=$(losetup -P -f --show "${image}")
 
 	#Refresh partition table
-	partprobe -s "${DEVICE}"
+	partprobe -s "${device}" > /dev/null
 	# Poll for the block device p1
-	local i=0
-	local max_tries=5
-	while [[ "$i" < "$max_tries" ]]; do
-		[ -b "${DEVICE}p1" ] && break
-		((i+=1))
-		echo -n "."
+	for _ in $(seq 1 5); do
+		if [ -b "${device}p1" ]; then
+			echo "${device}"
+			return 0
+		fi
 		sleep 1
 	done
-	[[ "$i" == "$max_tries" ]] && die "File ${DEVICE}p1 is not a block device"
 
-	MOUNT_DIR=$(mktemp -d osbuilder-mount-dir.XXXX)
-	info "Formatting Image using ext4 filesystem"
-	mkfs.ext4 -q -F -b "${BLOCK_SIZE}" "${DEVICE}p1"
-	OK "Image formatted"
+	error "File ${device}p1 is not a block device"
+	return 1
+}
+
+format_loop() {
+	local device="$1"
+	local block_size="$2"
+
+	case "${fs_type}" in
+		"${ext4_format}")
+			mkfs.ext4 -q -F -b "${block_size}" "${device}p1"
+			info "Set filesystem reserved blocks percentage to ${reserved_blocks_percentage}%"
+			tune2fs -m "${reserved_blocks_percentage}" "${device}p1"
+			;;
+
+		"${xfs_format}")
+			mkfs.xfs -q -f -b size="${block_size}" "${device}p1"
+			;;
+
+		*)
+			error "Unsupported fs type: ${fs_type}"
+			return 1
+			;;
+	esac
+}
+
+create_disk() {
+	local image="$1"
+	local img_size="$2"
+	local fs_type="$3"
+	local part_start="$4"
+
+	info "Creating raw disk with size ${img_size}M"
+	qemu-img create -q -f raw "${image}" "${img_size}M"
+	OK "Image file created"
+
+	# Kata runtime expect an image with just one partition
+	# The partition is the rootfs content
+	info "Creating partitions"
+	parted -s -a optimal "${image}" -- \
+		   mklabel msdos \
```

Comment:
So we are moving from gpt to msdos, the kernel is not able to read if gpt is used? this makes me think that should be good to have tested nemu as well.

Reply:
mbr is smaller than gpt, that way DAX metadata can be set

Reply:
kernel is in charge to read the partition table, not the hypervisor

Reply:
I mention because last I think I know was that ovmf  as firmware not sure if mbr vs gpt could could bug something there @sboeuf  you may be more updated on that topic